From 853bd291f02fbd1c6facd6d9b2f35c8d3bf46052 Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Thu, 24 Oct 2024 17:22:13 +0800
Subject: [PATCH] WLANRTSYS-85867 Drop the Assoc Req if the previous Assoc req
 is pending for the same STA

Issue:
	When STA sends assoc req again before the completion of previous
	assoc response callback, GroupId 14 fatal is observed.
Fix:
	Drop the assoc request, if the the same Sta's previous assoc request
	callback is pending.

Fixup: WLANRTSYS-62845 MLO changes in iwlwav-hostap-ng: AP MLD Auth Assoc and eapol changes
---
 src/ap/ieee802_11.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 9c84288ab..4f59b99e6 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5860,6 +5860,11 @@ static void handle_assoc(struct hostapd_data *hapd,
 		wpa_printf(MSG_ERROR, "Sta " MACSTR " already in the connected list, drop assoc", MAC2STR(mgmt->sa));
 		return;
 	}
+
+	if (sta && (sta->flags & WLAN_STA_ASSOC_REQ_OK) && !(sta->flags & WLAN_STA_ASSOC)) {
+		wpa_printf(MSG_ERROR, "Sta " MACSTR "previous assoc req is pending, drop assoc", MAC2STR(mgmt->sa));
+		return;
+	}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
 #ifdef CONFIG_IEEE80211R_AP
-- 
2.43.0

